tests/interfaces/test_batch_autograd.py

```diff
@@ -0,0 +1,584 @@
+        """Test that an error is raised if the interface is unknown"""
+        a = np.array([0.1, 0.2], requires_grad=True)
+
+        dev = qml.device("default.qubit", wires=1)
+
+        def cost(a, device):
+            with qml.tape.JacobianTape() as tape:
+                qml.RY(a[0], wires=0)
+                qml.RX(a[1], wires=0)
+                qml.expval(qml.PauliZ(0))
+
+            return execute([tape], device, gradient_fn=param_shift, interface=None)[0]
+
+        with pytest.raises(ValueError, match="Unknown interface"):
+            cost(a, device=dev)
+
+    def test_forward_mode(self, mocker):
+        """Test that forward mode uses the `device.execute_and_gradients` pathway"""
+        dev = qml.device("default.qubit", wires=1)
+        spy = mocker.spy(dev, "execute_and_gradients")
+
+        def cost(a):
+            with qml.tape.JacobianTape() as tape:
+                qml.RY(a[0], wires=0)
+                qml.RX(a[1], wires=0)
+                qml.expval(qml.PauliZ(0))
+
+            return execute(
+                [tape],
+                dev,
+                gradient_fn="device",
+                gradient_kwargs={"method": "adjoint_jacobian", "use_device_state": True},
+            )[0]
+
+        a = np.array([0.1, 0.2], requires_grad=True)
+        cost(a)
+
+        # adjoint method only performs a single device execution, but gets both result and gradient
+        assert dev.num_executions == 1
+        spy.assert_called()
+
+    def test_backward_mode(self, mocker):
+        """Test that backward mode uses the `device.batch_execute` and `device.gradients` pathway"""
+        dev = qml.device("default.qubit", wires=1)
+        spy_execute = mocker.spy(qml.devices.DefaultQubit, "batch_execute")
+        spy_gradients = mocker.spy(qml.devices.DefaultQubit, "gradients")
+
+        def cost(a):
+            with qml.tape.JacobianTape() as tape:
+                qml.RY(a[0], wires=0)
+                qml.RX(a[1], wires=0)
+                qml.expval(qml.PauliZ(0))
+
+            return execute(
+                [tape],
+                dev,
+                gradient_fn="device",
+                mode="backward",
+                gradient_kwargs={"method": "adjoint_jacobian"},
+            )[0]
+
+        a = np.array([0.1, 0.2], requires_grad=True)
+        cost(a)
+
+        assert dev.num_executions == 1
+        spy_execute.assert_called()
+        spy_gradients.assert_not_called()
+
+        qml.jacobian(cost)(a)
+        spy_gradients.assert_called()
+
+
+execute_kwargs = [
+    {"gradient_fn": param_shift},
+    {
+        "gradient_fn": "device",
+        "mode": "forward",
+        "gradient_kwargs": {"method": "adjoint_jacobian", "use_device_state": True},
+    },
+    {
+        "gradient_fn": "device",
+        "mode": "backward",
+        "gradient_kwargs": {"method": "adjoint_jacobian"},
+    },
+]
+
+
+@pytest.mark.parametrize("execute_kwargs", execute_kwargs)
+class TestAutogradExecuteIntegration:

```

Comment:
Note: This entire test class has been adapted directly from https://github.com/PennyLaneAI/pennylane/blob/master/tests/interfaces/test_tape_autograd.py to work with the new interface function.

Thus, the new `autograd.execute` function passes all the same integration tests as the current autograd interface.

Reply:
> Can we use the parent tape class instead for all these tests?

Not yet, but that is the plan once all PRs in this story are complete! Currently, there are a few auxiliary methods in the `JacobianTape` (namely, `JacobianTape._grad_method`, `JacobianTape._update_gradient_info`, and `JacobianTape._grad_method_validation`) that are needed for circuit gradients to work.

I actually realized this late -- I initially did try to write these tests using `qml.tape.QuantumTape()`, only to run into a lot of failures.

I still want to delete the `JacobianTape`, so there will just need to be a new ticket in this story to work out where to move the functionality currently in those class methods.

We could either move them to the parent class, or turn them into functions.